### ISSUE TITLE
Materializing a Set with a reference signals an ImproperOperation

### ISSUE THREAD

**Issue by Rinzwind** [GsDevKit/ston#11]
Materializing a Set with a reference signals an ImproperOperation in GemStone 3.2.17:

```
STON reader on: '[Point[0,0],Set[@2]]' readStream; next.
```
⇒ `a ImproperOperation occurred (error 2717), (instVarAt: 5 put:) not allowed, first public offset 6`

The error is signaled in the context of [`UnorderedCollection>>stonProcessSubObjects:`](https://github.com/GsDevKit/ston/blob/25aace18006bd8749a379741824e800d317977fb/repository/STON-GemStone-Core.package/UnorderedCollection.extension/instance/stonProcessSubObjects..st#L9).

**Comment by Rinzwind** [comment]
As a work-around, I've implemented `Set>>#stonProcessSubObjects:` as:

```
stonProcessSubObjects: block

	dict stonProcessSubObjects: block.
```

I'm not sure whether that is also sufficient as a fix. As for the other subclasses of `UnorderedCollection`, a similar work-around may be needed but I didn't look at those in detail as the STONs I need to materialize only involve `Set`.

**Comment by dalehenrich** [comment]
Good catch ... clearly the test suite for STON isn't covering all of the use cases properly ... The other subclasses that appear to be exposed to this bug are: RcIdenityBag, UserProfileGroup, UserProfileSet, and Bag .... so we'll need coverage for these classes (and bugfixes, presumably) ... if you could provide a pull request with your partial fix that would be great!

**Comment by Rinzwind** [comment]
I realized the above fix is incomplete in that the Set is not rehashed. I'm trying to produce a more complete fix. I ran into another related bug in STON though: https://github.com/svenvc/ston/issues/22

**Comment by svenvc** [comment]
Well, that works on Pharo

STON fromString: '[Point[0,0],Set[@2]]'.

And indeed, rehashing is needed, see #stonPostReferenceResolution which also requires #isHealthy for efficiency.

I can imagine Gemstone having certain issues in that area (with large persistent collections)

**Comment by svenvc** [comment]
BTW, it is necessary for Gemstone to use the most recent STON version, as there were important changes; the GitHub repo is now the canonical one.

**Comment by Rinzwind** [comment]
Indeed, the [gemstone branch](https://github.com/GsDevKit/ston/tree/gemstone) seems a bit outdated with the latest commit (25aace1) dating back to 2015.

For now, I've copied over two tests from the main STON project repository (`#testSetWithReferenceElements` and `#testSetWithIndirectReferenceElements`) plus the minimum set of changes that I needed to make those work.

I implemented `#stonProcessSubObjects:` and `#stonPostReferenceResolution` a bit differently for GemStone: rather than first resolving the references "in-place" and then rehashing the Set if it has incorrectly hashed elements, [`#stonProcessSubObjects:`](https://github.com/Rinzwind/ston/blob/71ebfcfa5b91c228af644f641ab0658757309e14/repository/STON-GemStone-Core.package/Set.extension/instance/stonProcessSubObjects..st) simply always readds all elements to the Set while resolving them, and `#stonPostReferenceResolution` does nothing (there's no equivalent to Pharo's `#isHealthy` in GemStone as far as I know). I considered using the approach with a “placeholder” collection which I described in 
https://github.com/svenvc/ston/issues/22#issuecomment-453562552, but I couldn't get that to fully work yet.
